### PR TITLE
Add a tool for grabbing and parsing prometheus metrics

### PR DESCRIPTION
**Pull request by gmarek**
It's sad that it looks like this...

cc @wojtek-t

### CODE REVIEW

**Comment by gmarek** [comment]
@wojtek-t PTAL

File: pkg/metrics/api_server_metrics.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+import (
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/prometheus/common/model"
+)
+
+var KnownApiServerMetrics = map[string][]string{
+	"apiserver_request_count":                        {"verb", "resource", "client", "code"},
+	"apiserver_request_latencies_bucket":             {"verb", "resource", "le"},
```
Comment (wojtek-t) [inline pkg/metrics/api_server_metrics.go]: what is "le" ?

File: pkg/metrics/api_server_metrics.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	result := NewMetrics()
+	for metric := range KnownApiServerMetrics {
+		result[metric] = make(model.Samples, 0)
+	}
+	return ApiServerMetrics(result)
+}
+
+func parseApiServerMetrics(data string, unknownMetrics sets.String) (ApiServerMetrics, error) {
+	result := NewApiServerMetrics()
+	err := parseMetrics(data, KnownApiServerMetrics, (*Metrics)(&result), unknownMetrics)
```
Comment (wojtek-t) [inline pkg/metrics/api_server_metrics.go]: if err = parseMetrics(...); err != nil {
  return ...
}

File: pkg/metrics/api_server_metrics.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+	if err != nil {
+		return ApiServerMetrics{}, err
+	}
+	return result, nil
+}
+
+func (g *MetricsGrabber) getMetricsFromApiServer() (string, error) {
+	if rawOutput, err := g.client.Get().RequestURI("/metrics").Do().Raw(); err != nil {
+		return "", err
+	} else {
```
Comment (wojtek-t) [inline pkg/metrics/api_server_metrics.go]: Can you remove "else" here.

File: pkg/metrics/controller_manager_metrics.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	result := NewMetrics()
+	for metric := range KnownControllerManagerMetrics {
+		result[metric] = make(model.Samples, 0)
+	}
+	return ControllerManagerMetrics(result)
+}
+
+func parseControllerManagerMetrics(data string, unknownMetrics sets.String) (ControllerManagerMetrics, error) {
+	result := NewControllerManagerMetrics()
+	err := parseMetrics(data, KnownControllerManagerMetrics, (*Metrics)(&result), unknownMetrics)
```
Comment (wojtek-t) [inline pkg/metrics/controller_manager_metrics.go]: if err = parseMetrics(...); err != nil {
return ...
}

File: pkg/metrics/api_server_metrics.go (hunk 1 above)
Comment (gmarek) [inline pkg/metrics/api_server_metrics.go]: It's an existing label "lesser equal"

File: pkg/metrics/api_server_metrics.go (hunk 2 above)
Comment (gmarek) [inline pkg/metrics/api_server_metrics.go]: Done.

File: pkg/metrics/api_server_metrics.go (hunk 3 above)
Comment (gmarek) [inline pkg/metrics/api_server_metrics.go]: Done.

File: pkg/metrics/controller_manager_metrics.go (hunk 4 above)
Comment (gmarek) [inline pkg/metrics/controller_manager_metrics.go]: Done.

File: pkg/metrics/kubelet_metrics.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	result := NewMetrics()
+	for metric := range KnownKubeletMetrics {
+		result[metric] = make(model.Samples, 0)
+	}
+	return KubeletMetrics(result)
+}
+
+func parseKubeletMetrics(data string, unknownMetrics sets.String) (KubeletMetrics, error) {
+	result := NewKubeletMetrics()
+	err := parseMetrics(data, KnownKubeletMetrics, (*Metrics)(&result), unknownMetrics)
```
Comment (wojtek-t) [inline pkg/metrics/kubelet_metrics.go]: if err = parseMetrics(...); err != nil {
return ...
}

File: pkg/metrics/scheduler_metrics.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+	result := NewMetrics()
+	for metric := range KnownSchedulerMetrics {
+		result[metric] = make(model.Samples, 0)
+	}
+	return SchedulerMetrics(result)
+}
+
+func parseSchedulerMetrics(data string, unknownMetrics sets.String) (SchedulerMetrics, error) {
+	result := NewSchedulerMetrics()
+	err := parseMetrics(data, KnownSchedulerMetrics, (*Metrics)(&result), unknownMetrics)
```
Comment (wojtek-t) [inline pkg/metrics/scheduler_metrics.go]: if err = parseMetrics(...); err != nil {
return ...
}

File: pkg/metrics/metrics_grabber.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+	SchedulerMetrics         SchedulerMetrics
+}
+
+type MetricsGrabber struct {
+	client                    *client.Client
+	grabFromApiServer         bool
+	grabFromControllerManager bool
+	grabFromKubelets          bool
+	grabFromScheduler         bool
+	kubeletPort               int
```
Comment (wojtek-t) [inline pkg/metrics/metrics_grabber.go]: What if kubelets have different port? Can't we use data read from apiserver?

File: pkg/metrics/metrics_grabber.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+func NewMetricsGrabber(c *client.Client, kubelets bool, scheduler bool, controllers bool, apiServer bool) (*MetricsGrabber, error) {
+	kubeletPort := 0
+	registeredMaster := false
+	masterName := ""
+	nodeList, err := c.Nodes().List(api.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	if len(nodeList.Items) < 1 {
+		return nil, fmt.Errorf("Can't find any Nodes in the API server to grab metrics from")
```
Comment (wojtek-t) [inline pkg/metrics/metrics_grabber.go]: I can imagine a case that all kubelets died, but we still want to grab metrics from master components.

File: pkg/metrics/metrics_grabber.go [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	for _, node := range nodeList.Items {
+		if strings.HasSuffix(node.Name, "master") {
```
Comment (wojtek-t) [inline pkg/metrics/metrics_grabber.go]: This is a hacky solution. Can you please create a separate function like isMasterNode and leave a TODO there to make it less hacky :)

File: pkg/metrics/metrics_grabber.go [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+			registeredMaster = true
+			masterName = node.Name
+			if kubeletPort != 0 {
```
Comment (wojtek-t) [inline pkg/metrics/metrics_grabber.go]: This is over-optimization in my opinion that is unnecessary complicating the code.

File: pkg/metrics/metrics_grabber.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+				break
+			}
+		} else {
+			// We assume that all Kubelets run on the same port. This will need to get change if assumption is prover wrong.
+			kubeletPort = node.Status.DaemonEndpoints.KubeletEndpoint.Port
+			if registeredMaster {
```
Comment (wojtek-t) [inline pkg/metrics/metrics_grabber.go]: This is over-optimization in my opinion that is unnecessary complicating the code.

File: pkg/metrics/metrics_grabber.go [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+	}
+	for _, node := range nodeList.Items {
+		if strings.HasSuffix(node.Name, "master") {
+			registeredMaster = true
+			masterName = node.Name
+			if kubeletPort != 0 {
+				break
+			}
+		} else {
+			// We assume that all Kubelets run on the same port. This will need to get change if assumption is prover wrong.
```
Comment (wojtek-t) [inline pkg/metrics/metrics_grabber.go]: nit: can you please split this line?

File: pkg/metrics/metrics_grabber.go (hunk 7 above)
Comment (gmarek) [inline pkg/metrics/metrics_grabber.go]: We are - see NewMetricsGrabber.

File: pkg/metrics/metrics_grabber.go [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+			kubeletPort = node.Status.DaemonEndpoints.KubeletEndpoint.Port
+			if registeredMaster {
+				break
+			}
+		}
+	}
+	if !registeredMaster {
+		glog.Warningf("Master node is not registered. Grabbing metrics from Scheduler, ControllerManager and API server is disabled.")
```
Comment (wojtek-t) [inline pkg/metrics/metrics_grabber.go]: shouldnt' you overwrite: apiServer, controllers, scheduler to false here?

File: test/e2e/metrics_grabber_test.go [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+		for _, label := range labels {
+			unobservedLabels.Insert(label)
+		}
+		vector, found := data[metric]
+		Expect(found).To(Equal(true))
+		if len(vector) == 0 {
+			Logf("Missing: %v", metric)
+			missing.Insert(metric)
+		} else {
+			if metric == "kubelet_generate_pod_status_latency_microsecond" {
```
Comment (wojtek-t) [inline test/e2e/metrics_grabber_test.go]: please remove

File: test/e2e/metrics_grabber_test.go [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+// Missing = Assumed minus Observed, Invalid = Observed minus Assumed
+func validateLabelSet(labelSet map[string][]string, data metrics.Metrics, missing sets.String, invalidLabels map[string]sets.String, missingLabels map[string]sets.String) {
+	for metric, labels := range labelSet {
+		unobservedLabels := sets.NewString()
+		for _, label := range labels {
+			unobservedLabels.Insert(label)
+		}
+		vector, found := data[metric]
+		Expect(found).To(Equal(true))
+		if len(vector) == 0 {
```
Comment (wojtek-t) [inline test/e2e/metrics_grabber_test.go]: if !found || len(vector) == 0 
?

File: test/e2e/metrics_grabber_test.go [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+			if metric == "kubelet_generate_pod_status_latency_microsecond" {
+				Logf("Found: %v", vector)
+			}
+			for _, observation := range vector {
+				for label := range observation.Metric {
+					// When we notice some label, remove it from the unobserved ones
+					unobservedLabels.Delete(string(label))
+
+					// We need to check if it's a known label for this metric.
+					// Omit Promethesu internal metrics.
```
Comment (wojtek-t) [inline test/e2e/metrics_grabber_test.go]: s/Promethesu/Prometheus/

**Comment by wojtek-t** [comment]
I added some comments - but those are mostly minor ones.

File: pkg/metrics/metrics_grabber.go (hunk 9 above)
Comment (gmarek) [inline pkg/metrics/metrics_grabber.go]: Done.

File: pkg/metrics/metrics_grabber.go (hunk 8 above)
Comment (gmarek) [inline pkg/metrics/metrics_grabber.go]: Done.

File: pkg/metrics/metrics_grabber.go (hunk 10 above)
Comment (gmarek) [inline pkg/metrics/metrics_grabber.go]: Done.

File: pkg/metrics/metrics_grabber.go (hunk 11 above)
Comment (gmarek) [inline pkg/metrics/metrics_grabber.go]: Done.

File: pkg/metrics/metrics_grabber.go (hunk 12 above)
Comment (gmarek) [inline pkg/metrics/metrics_grabber.go]: Done.

File: pkg/metrics/metrics_grabber.go (hunk 13 above)
Comment (gmarek) [inline pkg/metrics/metrics_grabber.go]: We're checking it below, but it's probably cleaner here. Thanks.

File: test/e2e/metrics_grabber_test.go (hunk 14 above)
Comment (gmarek) [inline test/e2e/metrics_grabber_test.go]: Ups:)

File: test/e2e/metrics_grabber_test.go (hunk 15 above)
Comment (gmarek) [inline test/e2e/metrics_grabber_test.go]: It's impossible to get !found, and we actually check that in the line above.

Comment (wojtek-t) [inline test/e2e/metrics_grabber_test.go]: Failed Expect will just log an error, not stop the function - it will still be running

Comment (gmarek) [inline test/e2e/metrics_grabber_test.go]: OK - it's impossible because we're creating all keys in the output metrics. No harm in adding this check though.

**Comment by gmarek** [comment]
@k8s-bot test this

@wojtek-t PTAL

**Comment by wojtek-t** [comment]
This test failed previously:
"Kubernetes e2e suite.MetricsGrabber should grab all metrics from API server."

@gmarek Can you please take a look?

I started wondering if we should have such a strict test...

**Comment by gmarek** [comment]
@wojtek-t Done.

**Comment by wojtek-t** [comment]
LGTM

**Comment by gmarek** [comment]
Failure is fixed by #18917

@k8s-bot unit test this

**Pull request merged**